File: lib/es6/belt_MutableSet.mjs


import * as Curry from "./curry.mjs";
import * as Belt_SortArray from "./belt_SortArray.mjs";
import * as Belt_internalAVLset from "./belt_internalAVLset.mjs";

function remove0(nt, x, cmp) {
  var k = nt.v;
  var c = cmp(x, k);
  if (c === 0) {
    var l = nt.l;
    var r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (c < 0) {
    var l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = remove0(l$1, x, cmp);
      return Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  var r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = remove0(r$1, x, cmp);
    return Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function remove(d, v) {
  var oldRoot = d.data;
  if (oldRoot === undefined) {
    return ;
  }
  var newRoot = remove0(oldRoot, v, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return ;
  }
  
}

function removeMany0(_t, xs, _i, len, cmp) {
  while(true) {
    var i = _i;
    var t = _t;
    if (i >= len) {
      return t;
    }
    var ele = xs[i];
    var u = remove0(t, ele, cmp);
    if (u === undefined) {
      return ;
    }
    _i = i + 1 | 0;
    _t = u;
    continue ;
  };
}

function removeMany(d, xs) {
  var oldRoot = d.data;
  if (oldRoot === undefined) {
    return ;
  }
  var len = xs.length;
  d.data = removeMany0(oldRoot, xs, 0, len, d.cmp);
  
}

function removeCheck0(nt, x, removed, cmp) {
  var k = nt.v;
  var c = cmp(x, k);
  if (c === 0) {
    removed.contents = true;
    var l = nt.l;
    var r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (c < 0) {
    var l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = removeCheck0(l$1, x, removed, cmp);
      return Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  var r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = removeCheck0(r$1, x, removed, cmp);
    return Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function removeCheck(d, v) {
  var oldRoot = d.data;
  if (oldRoot === undefined) {
    return false;
  }
  var removed = {
    contents: false
  };
  var newRoot = removeCheck0(oldRoot, v, removed, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
  }
  return removed.contents;
}

function addCheck0(t, x, added, cmp) {
  if (t !== undefined) {
    var k = t.v;
    var c = cmp(x, k);
    if (c === 0) {
      return t;
    }
    var l = t.l;
    var r = t.r;
    if (c < 0) {
      var ll = addCheck0(l, x, added, cmp);
      t.l = ll;
    } else {
      t.r = addCheck0(r, x, added, cmp);
    }
    return Belt_internalAVLset.balMutate(t);
  }
  added.contents = true;
  return Belt_internalAVLset.singleton(x);
}

function addCheck(m, e) {
  var oldRoot = m.data;
  var added = {
    contents: false
  };
  var newRoot = addCheck0(oldRoot, e, added, m.cmp);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
  }
  return added.contents;
}

function add(m, e) {
  var oldRoot = m.data;
  var newRoot = Belt_internalAVLset.addMutate(m.cmp, oldRoot, e);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
    return ;
  }
  
}

function addArrayMutate(t, xs, cmp) {
  var v = t;
  for(var i = 0 ,i_finish = xs.length; i < i_finish; ++i){
    v = Belt_internalAVLset.addMutate(cmp, v, xs[i]);
  }
  return v;
}

function mergeMany(d, xs) {
  d.data = addArrayMutate(d.data, xs, d.cmp);
  
}

function make(id) {
  return {
          cmp: id.cmp,
          data: undefined
        };
}

function isEmpty(d) {
  var n = d.data;
  return n === undefined;
}

function minimum(d) {
  return Belt_internalAVLset.minimum(d.data);
}

function minUndefined(d) {
  return Belt_internalAVLset.minUndefined(d.data);
}

function maximum(d) {
  return Belt_internalAVLset.maximum(d.data);
}

function maxUndefined(d) {
  return Belt_internalAVLset.maxUndefined(d.data);
}

function forEachU(d, f) {
  return Belt_internalAVLset.forEachU(d.data, f);
}

function forEach(d, f) {
  return Belt_internalAVLset.forEachU(d.data, Curry.__1(f));
}

function reduceU(d, acc, cb) {
  return Belt_internalAVLset.reduceU(d.data, acc, cb);
}

function reduce(d, acc, cb) {
  return reduceU(d, acc, Curry.__2(cb));
}

function everyU(d, p) {
  return Belt_internalAVLset.everyU(d.data, p);
}

function every(d, p) {
  return Belt_internalAVLset.everyU(d.data, Curry.__1(p));
}

function someU(d, p) {
  return Belt_internalAVLset.someU(d.data, p);
}

function some(d, p) {
  return Belt_internalAVLset.someU(d.data, Curry.__1(p));
}

function size(d) {
  return Belt_internalAVLset.size(d.data);
}

function toList(d) {
  return Belt_internalAVLset.toList(d.data);
}

function toArray(d) {
  return Belt_internalAVLset.toArray(d.data);
}

function fromSortedArrayUnsafe(xs, id) {
  return {
          cmp: id.cmp,
          data: Belt_internalAVLset.fromSortedArrayUnsafe(xs)
        };
}

function checkInvariantInternal(d) {
  return Belt_internalAVLset.checkInvariantInternal(d.data);
}

function fromArray(data, id) {
  var cmp = id.cmp;
  return {
          cmp: cmp,
          data: Belt_internalAVLset.fromArray(data, cmp)
        };
}

function cmp(d0, d1) {
  return Belt_internalAVLset.cmp(d0.data, d1.data, d0.cmp);
}

function eq(d0, d1) {
  return Belt_internalAVLset.eq(d0.data, d1.data, d0.cmp);
}

function get(d, x) {
  return Belt_internalAVLset.get(d.data, x, d.cmp);
}

function getUndefined(d, x) {
  return Belt_internalAVLset.getUndefined(d.data, x, d.cmp);
}

function getExn(d, x) {
  return Belt_internalAVLset.getExn(d.data, x, d.cmp);
}

function split(d, key) {
  var arr = Belt_internalAVLset.toArray(d.data);
  var cmp = d.cmp;
  var i = Belt_SortArray.binarySearchByU(arr, key, cmp);
  var len = arr.length;
  if (i >= 0) {
    return [
            [
              {
                cmp: cmp,
                data: Belt_internalAVLset.fromSortedArrayAux(arr, 0, i)
              },
              {
                cmp: cmp,
                data: Belt_internalAVLset.fromSortedArrayAux(arr, i + 1 | 0, (len - i | 0) - 1 | 0)
              }
            ],
            true
          ];
  }
  var next = (-i | 0) - 1 | 0;
  return [
          [
            {
              cmp: cmp,
              data: Belt_internalAVLset.fromSortedArrayAux(arr, 0, next)
            },
            {
              cmp: cmp,
              data: Belt_internalAVLset.fromSortedArrayAux(arr, next, len - next | 0)
            }
          ],
          false
        ];
}

function keepU(d, p) {
  return {
          cmp: d.cmp,
          data: Belt_internalAVLset.keepCopyU(d.data, p)
        };
}

function keep(d, p) {
  return keepU(d, Curry.__1(p));
}

function partitionU(d, p) {
  var cmp = d.cmp;
  var match = Belt_internalAVLset.partitionCopyU(d.data, p);
  return [
          {
            cmp: cmp,
            data: match[0]
          },
          {
            cmp: cmp,
            data: match[1]
          }
        ];
}

function partition(d, p) {
  return partitionU(d, Curry.__1(p));
}

function subset(a, b) {
  return Belt_internalAVLset.subset(a.data, b.data, a.cmp);
}

function intersect(a, b) {
  var cmp = a.cmp;
  var match = a.data;
  var match$1 = b.data;
  if (match === undefined) {
    return {
            cmp: cmp,
            data: undefined
          };
  }
  if (match$1 === undefined) {
    return {
            cmp: cmp,
            data: undefined
          };
  }
  var sizea = Belt_internalAVLset.lengthNode(match);
  var sizeb = Belt_internalAVLset.lengthNode(match$1);
  var totalSize = sizea + sizeb | 0;
  var tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(match, 0, tmp);
  Belt_internalAVLset.fillArray(match$1, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || cmp(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
    return {
            cmp: cmp,
            data: undefined
          };
  }
  var tmp2 = new Array(sizea < sizeb ? sizea : sizeb);
  var k = Belt_SortArray.intersectU(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
          cmp: cmp,
          data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
        };
}

function diff(a, b) {
  var cmp = a.cmp;
  var dataa = a.data;
  var match = b.data;
  if (dataa === undefined) {
    return {
            cmp: cmp,
            data: undefined
          };
  }
  if (match === undefined) {
    return {
            cmp: cmp,
            data: Belt_internalAVLset.copy(dataa)
          };
  }
  var sizea = Belt_internalAVLset.lengthNode(dataa);
  var sizeb = Belt_internalAVLset.lengthNode(match);
  var totalSize = sizea + sizeb | 0;
  var tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa, 0, tmp);
  Belt_internalAVLset.fillArray(match, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || cmp(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
    return {
            cmp: cmp,
            data: Belt_internalAVLset.copy(dataa)
          };
  }
  var tmp2 = new Array(sizea);
  var k = Belt_SortArray.diffU(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
          cmp: cmp,
          data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
        };
}

function union(a, b) {
  var cmp = a.cmp;
  var dataa = a.data;
  var datab = b.data;
  if (dataa === undefined) {
    return {
            cmp: cmp,
            data: Belt_internalAVLset.copy(datab)
          };
  }
  if (datab === undefined) {
    return {
            cmp: cmp,
            data: Belt_internalAVLset.copy(dataa)
          };
  }
  var sizea = Belt_internalAVLset.lengthNode(dataa);
  var sizeb = Belt_internalAVLset.lengthNode(datab);
  var totalSize = sizea + sizeb | 0;
  var tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa, 0, tmp);
  Belt_internalAVLset.fillArray(datab, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0) {
    return {
            cmp: cmp,
            data: Belt_internalAVLset.fromSortedArrayAux(tmp, 0, totalSize)
          };
  }
  var tmp2 = new Array(totalSize);
  var k = Belt_SortArray.unionU(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
          cmp: cmp,
          data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
        };
}

function has(d, x) {
  return Belt_internalAVLset.has(d.data, x, d.cmp);
}

function copy(d) {
  return {
          cmp: d.cmp,
          data: Belt_internalAVLset.copy(d.data)
        };
}

var Int;

var $$String;

export {
  Int ,
  $$String ,
  make ,
  fromArray ,
  fromSortedArrayUnsafe ,
  copy ,
  isEmpty ,
  has ,
  add ,
  addCheck ,
  mergeMany ,
  remove ,
  removeCheck ,
  removeMany ,
  union ,
  intersect ,
  diff ,
  subset ,
  cmp ,
  eq ,
  forEachU ,
  forEach ,
  reduceU ,
  reduce ,
  everyU ,
  every ,
  someU ,
  some ,
  keepU ,
  keep ,
  partitionU ,
  partition ,
  size ,
  toList ,
  toArray ,
  minimum ,
  minUndefined ,
  maximum ,
  maxUndefined ,
  get ,
  getUndefined ,
  getExn ,
  split ,
  checkInvariantInternal ,
  
}
/* No side effect */

File: syntax/tests/idempotency/bucklescript/belt/belt_MutableSet.ml

(* Copyright (C) 2017 Authors of BuckleScript
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module Int = Belt_MutableSetInt
module String = Belt_MutableSetString

module N = Belt_internalAVLset
module A = Belt_Array
module Sort = Belt_SortArray


type ('k, 'id) id = ('k, 'id) Belt_Id.comparable
type ('key, 'id ) cmp = ('key, 'id) Belt_Id.cmp

module S = struct
  type ('value,'id) t =
    {
      cmp: ('value, 'id) cmp;
      mutable data: 'value N.t
    } [@@bs.deriving abstract]
end

type ('k, 'id) t = ('k, 'id) S.t


let rec remove0 nt x ~cmp =
  let k = N.valueGet nt in
  let c = cmp x k [@bs] in
  if c = 0 then
    let l,r = N.(leftGet nt, rightGet nt) in
    match N.(toOpt l, toOpt r) with
    | None, _ -> r
    | _, None -> l
    | Some _,  Some nr ->
      N.rightSet nt (N.removeMinAuxWithRootMutate nt nr);
      N.return (N.balMutate nt)
   else
    begin
      if c < 0 then
        match N.toOpt (N.leftGet nt) with
        | None -> N.return nt
        | Some l ->
          N.leftSet nt (remove0 ~cmp l x );
          N.return (N.balMutate nt)
      else
        match N.toOpt (N.rightGet nt) with
        | None -> N.return nt
        | Some r ->
          N.rightSet nt (remove0 ~cmp r x);
          N.return (N.balMutate nt)
    end

let remove  d  v =
  let oldRoot = S.dataGet d in
  match N.toOpt oldRoot with
  | None -> ()
  | Some oldRoot2 ->
    let newRoot = remove0 ~cmp:(Belt_Id.getCmpInternal (S.cmpGet d)) oldRoot2 v in
    if newRoot != oldRoot then
      S.dataSet d newRoot


let rec removeMany0 t xs i len ~cmp  =
  if i < len then
    let ele = A.getUnsafe xs i in
    let u = remove0 t ele ~cmp in
    match N.toOpt u with
    | None -> N.empty
    | Some t -> removeMany0 t xs (i+1) len ~cmp
  else N.return t

let removeMany d xs =
  let oldRoot = S.dataGet d in
  match N.toOpt oldRoot with
  | None -> ()
  | Some nt ->
    let len = A.length xs in
    S.dataSet d
      (removeMany0 nt xs 0 len
        ~cmp:(Belt_Id.getCmpInternal (S.cmpGet d)))


let rec removeCheck0  nt x removed ~cmp=
  let k = N.valueGet nt in
  let c = (Belt_Id.getCmpInternal cmp) x k [@bs] in
  if c = 0 then
    let () = removed .contents<- true in
    let l,r = N.(leftGet nt, rightGet nt) in
    match N.(toOpt l, toOpt r) with
    | None, _ -> r
    | _, None -> l
    | Some _,  Some nr ->
      N.rightSet nt (N.removeMinAuxWithRootMutate nt nr);
      N.return (N.balMutate nt)
  else
    begin
      if c < 0 then
        match N.toOpt (N.leftGet nt) with
        | None -> N.return nt
        | Some l ->
          N.leftSet nt (removeCheck0 ~cmp l x removed);
          N.return (N.balMutate nt)
      else
        match N.toOpt (N.rightGet nt) with
        | None -> N.return nt
        | Some r ->
          N.rightSet nt (removeCheck0 ~cmp r x removed);
          N.return (N.balMutate nt)
    end



let removeCheck d v =
  let oldRoot = S.dataGet d in
  match N.toOpt oldRoot with
  | None -> false
  | Some oldRoot2 ->
    let removed = ref false in
    let newRoot = removeCheck0 ~cmp:(S.cmpGet d) oldRoot2 v removed in
    if newRoot != oldRoot then
      S.dataSet d newRoot ;
    removed.contents



let rec addCheck0  t x added ~cmp  =
  match N.toOpt t with
  | None ->
    added .contents<- true;
    N.singleton x
  | Some nt ->
    let k = N.valueGet nt in
    let c = cmp x k [@bs] in
    if c = 0 then t
    else
      let l, r = N.(leftGet nt, rightGet nt) in
      (if c < 0 then
         let ll = addCheck0 ~cmp l x added in
         N.leftSet nt ll
       else
         N.rightSet nt (addCheck0 ~cmp r x added );
      );
      N.return (N.balMutate nt)

let addCheck m e =
  let oldRoot = S.dataGet m in
  let added = ref false in
  let newRoot = addCheck0 ~cmp:(Belt_Id.getCmpInternal (S.cmpGet m)) oldRoot e added in
  if newRoot != oldRoot then
    S.dataSet m newRoot;
  added.contents

let add m e =
  let oldRoot = S.dataGet m in
  let newRoot = N.addMutate ~cmp:(S.cmpGet m) oldRoot e  in
  if newRoot != oldRoot then
    S.dataSet m newRoot

let addArrayMutate t xs ~cmp =
  let v = ref t in
  for i = 0 to A.length xs - 1 do
    v .contents<- N.addMutate v.contents (A.getUnsafe xs i)  ~cmp
  done;
  v.contents

let mergeMany d xs =
  S.dataSet d (addArrayMutate (S.dataGet d) xs ~cmp:(S.cmpGet d))


let make (type value) (type identity) ~(id : (value, identity) id) =
  let module M = (val id) in
  S.t ~cmp:M.cmp ~data:N.empty

let isEmpty d =
  N.isEmpty (S.dataGet d)

let minimum d =
  N.minimum (S.dataGet d)
let minUndefined d =
  N.minUndefined (S.dataGet d)
let maximum d =
  N.maximum (S.dataGet d)
let maxUndefined d =
  N.maxUndefined (S.dataGet d)

let forEachU d f = N.forEachU (S.dataGet d) f
let forEach d f = forEachU d (fun[@bs] a -> f a)
let reduceU d acc cb = N.reduceU (S.dataGet d) acc cb
let reduce d acc cb = reduceU d acc (fun[@bs] a b -> cb a b)
let everyU d p = N.everyU (S.dataGet d) p
let every d p = everyU d (fun[@bs] a -> p a)
let someU d p = N.someU (S.dataGet d) p
let some d p = someU d (fun[@bs] a -> p a)
let size d =
  N.size (S.dataGet d)
let toList d =
  N.toList (S.dataGet d)
let toArray d =
  N.toArray (S.dataGet d)

let fromSortedArrayUnsafe (type value) (type identity) xs ~(id : (value,identity) id) : _ t =
  let module M = (val id) in
  S.t ~data:(N.fromSortedArrayUnsafe xs) ~cmp:M.cmp

let checkInvariantInternal d =
  N.checkInvariantInternal (S.dataGet d)


let fromArray (type value) (type identity)  data ~(id : (value,identity) id) =
  let module M = (val id) in
  let cmp = M.cmp in
  S.t ~cmp ~data:(N.fromArray ~cmp data)


let cmp d0 d1 =
  N.cmp ~cmp:(S.cmpGet d0) (S.dataGet d0) (S.dataGet d1)

let eq d0  d1 =
  N.eq ~cmp:(S.cmpGet d0) (S.dataGet d0) (S.dataGet d1)

let get d x =
  N.get ~cmp:(S.cmpGet d) (S.dataGet d) x

let getUndefined  d x =
  N.getUndefined ~cmp:(S.cmpGet d) (S.dataGet d) x

let getExn d x =
  N.getExn ~cmp:(S.cmpGet d) (S.dataGet d) x


let split d  key  =
  let arr = N.toArray (S.dataGet d) in
  let cmp = S.cmpGet d in
  let i = Sort.binarySearchByU arr key (Belt_Id.getCmpInternal cmp)  in
  let len = A.length arr in
  if i < 0 then
    let next = - i -1 in
    (S.t
       ~data:(N.fromSortedArrayAux arr 0 next)
       ~cmp
     ,
     S.t
       ~data:(N.fromSortedArrayAux arr next (len - next))
       ~cmp
    ), false
  else
    (S.t
       ~data:(N.fromSortedArrayAux arr 0 i)
       ~cmp,
     S.t
       ~data:(N.fromSortedArrayAux arr (i+1) (len - i - 1))
       ~cmp
    ), true

let keepU d p =
  S.t ~data:(N.keepCopyU (S.dataGet d) p ) ~cmp:(S.cmpGet d)

let keep d p = keepU d (fun[@bs] a -> p a)

let partitionU d p =
  let cmp = S.cmpGet d in
  let a, b = N.partitionCopyU (S.dataGet d) p in
  S.t ~data:a ~cmp, S.t ~data:b ~cmp

let partition d p = partitionU d (fun[@bs] a -> p a)

let subset a b =
  N.subset  ~cmp:(S.cmpGet a) (S.dataGet a) (S.dataGet b)

let intersect a b  : _ t =
  let cmp = S.cmpGet a  in
  match N.toOpt (S.dataGet a), N.toOpt (S.dataGet b) with
  | None, _ -> S.t ~cmp ~data:N.empty
  | _, None -> S.t ~cmp ~data:N.empty
  | Some dataa0, Some datab0 ->
    let sizea, sizeb =
      N.lengthNode dataa0, N.lengthNode datab0 in
    let totalSize = sizea + sizeb in
    let tmp = A.makeUninitializedUnsafe totalSize in
    ignore (N.fillArray dataa0 0 tmp) ;
    ignore (N.fillArray datab0 sizea tmp);
    let p = Belt_Id.getCmpInternal cmp in
    if (p (A.getUnsafe tmp (sizea - 1))
          (A.getUnsafe tmp sizea) [@bs] < 0)
       ||
       (p
          (A.getUnsafe tmp (totalSize - 1))
          (A.getUnsafe tmp 0) [@bs] < 0
       )
    then S.t ~cmp ~data:N.empty
    else
      let tmp2 = A.makeUninitializedUnsafe (Pervasives.min sizea sizeb) in
      let k = Sort.intersectU tmp 0 sizea tmp sizea sizeb tmp2 0 p in
      S.t ~data:(N.fromSortedArrayAux tmp2 0 k)
        ~cmp

let diff a b : _ t =
  let cmp = S.cmpGet a in
  let dataa = S.dataGet a in
  match N.toOpt dataa, N.toOpt (S.dataGet b) with
  | None, _ -> S.t ~cmp ~data:N.empty
  | _, None ->
    S.t ~data:(N.copy dataa) ~cmp
  | Some dataa0, Some datab0
    ->
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in
    let totalSize = sizea + sizeb in
    let tmp = A.makeUninitializedUnsafe totalSize in
    ignore (N.fillArray dataa0 0 tmp) ;
    ignore (N.fillArray datab0 sizea tmp);
    let p = Belt_Id.getCmpInternal cmp in
    if (p (A.getUnsafe tmp (sizea - 1))
          (A.getUnsafe tmp sizea) [@bs] < 0)
       ||
       (p
          (A.getUnsafe tmp (totalSize - 1))
          (A.getUnsafe tmp 0) [@bs] < 0
       )
    then S.t ~data:(N.copy dataa) ~cmp
    else
      let tmp2 = A.makeUninitializedUnsafe sizea in
      let k = Sort.diffU tmp 0 sizea tmp sizea sizeb tmp2 0 p in
      S.t ~data:(N.fromSortedArrayAux tmp2 0 k) ~cmp

let union a b =
  let cmp = S.cmpGet a in
  let dataa, datab =  S.dataGet a, S.dataGet b  in
  match N.toOpt dataa, N.toOpt datab with
  | None, _ -> S.t ~data:(N.copy datab) ~cmp
  | _, None -> S.t ~data:(N.copy dataa) ~cmp
  | Some dataa0, Some datab0
    ->
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in
    let totalSize = sizea + sizeb in
    let tmp = A.makeUninitializedUnsafe totalSize in
    ignore (N.fillArray dataa0 0 tmp) ;
    ignore (N.fillArray datab0 sizea tmp) ;
    let p = (Belt_Id.getCmpInternal cmp)  in
    if p
        (A.getUnsafe tmp (sizea - 1))
        (A.getUnsafe tmp sizea) [@bs] < 0 then
      S.t ~data:(N.fromSortedArrayAux tmp 0 totalSize) ~cmp
    else
      let tmp2 = A.makeUninitializedUnsafe totalSize in
      let k = Sort.unionU tmp 0 sizea tmp sizea sizeb tmp2 0 p in
      S.t ~data:(N.fromSortedArrayAux tmp2 0 k) ~cmp

let has d x =
  N.has ~cmp:(S.cmpGet d) (S.dataGet d) x

let copy d = S.t ~data:(N.copy (S.dataGet d)) ~cmp:(S.cmpGet d)


